### PR TITLE
Fix a race condition in converting data layouts in MKLDNN.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
There is a race condition in data layout conversion in the MKLDNN implementation. 
Currently, when NDArray::data() is called, it converts data format (from the MKLDNN format to the default format) inside an NDArray. In the threaded execution engine, while an NDArray is being converted in a thread, the array can also be read by another thread. In this case, the other thread can read wrong data. A similar race condition may also exist in the MKLML integration, which sometimes generates the garbage results.

Please see the details of the error in https://github.com/apache/incubator-mxnet/issues/9820

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: src/ndarray/ndarray.cc [hunk 1]
```diff
@@ -1017,6 +1017,7 @@ inline void CopyFromToDnsImpl(const NDArray& from, const NDArray& to, RunContext
     // with Copy().
     NDArray tmp_from = from;
     if (tmp_from.IsMKLDNNData()) {
+      // TODO(zhengda) tmp_from should be cached.
```
Comment (marcoabreu) [inline src/ndarray/ndarray.cc]: Please create an issue to track this

File: tests/python/gpu/test_gluon_model_zoo_gpu.py [hunk 2]
```diff
@@ -37,11 +37,11 @@ def download_data():
     return mx.test_utils.download(
         'http://data.mxnet.io/data/val-5k-256.rec', VAL_DATA)
 
-@unittest.skip("test fails intermittently. temporarily disabled.")
 @with_seed()
 def test_inference():
     all_models = ['resnet50_v1', 'vgg19_bn', 'alexnet', #'inceptionv3',
                   'densenet201', 'squeezenet1.0', 'mobilenet0.25']
+    mx.random.seed(2)
```
Comment (marcoabreu) [inline tests/python/gpu/test_gluon_model_zoo_gpu.py]: please don't use mx.random.seed due to the introduction of the @with_seed() decorator

File: tests/python/gpu/test_gluon_model_zoo_gpu.py [hunk 3]
```diff
@@ -105,6 +107,7 @@ def test_training():
     # TODO(zhengda) mobilenet can't pass this test even without MKLDNN.
     all_models = ['resnet18_v1', 'densenet121']
 
+    mx.random.seed(1)
```
Comment (marcoabreu) [inline tests/python/gpu/test_gluon_model_zoo_gpu.py]: please don't use mx.random.seed due to the introduction of the @with_seed() decorator

File: tests/python/gpu/test_gluon_model_zoo_gpu.py (hunk 2 above)
Comment (zheng-da) [reply tests/python/gpu/test_gluon_model_zoo_gpu.py]: Does with_seed() set the same seed for each run?

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (zheng-da) [reply src/ndarray/ndarray.cc]: Create a github issue for this?

**Comment by zheng-da** [comment]
@marcoabreu previously, I disabled the inference test. Now I enabled all tests. I also added some prints to clearly see whether CPU or GPU generates wrong results.

File: src/operator/nn/mkldnn/mkldnn_base.cc [hunk 4]
```diff
@@ -270,9 +270,24 @@ void FallBackCompute(FCompute fn, const nnvm::NodeAttrs &attrs,
...
+  std::vector<NDArray> in_bufs;
   for (size_t i = 0; i < in_blobs.size(); i++) {
+    // If the input data isn't stored in the default format, we shouldn't
+    // call data() directly, which will change the layout of the NDArray.
+    // Instead, we should save the converted data in another NDArray.
+    // TODO(zhengda) we should use temp space to save the converted data.
+    if (inputs[i].IsDefaultData()) {
       in_blobs[i] = inputs[i].data();
+    } else {
+      in_bufs.emplace_back(inputs[i].shape(), inputs[i].ctx(),
```
Comment (cjolivier01) [inline src/operator/nn/mkldnn/mkldnn_base.cc]: nit: it would be nice to call in_bufs.reserve(in_blobs.size()) on this first pass (ie if in_bufs.empty())
If there is often more than one in_bufs to be inserted

File: src/operator/nn/mkldnn/mkldnn_base.cc [hunk 5]
```diff
@@ -270,9 +270,24 @@ void FallBackCompute(FCompute fn, const nnvm::NodeAttrs &attrs,
...
+                           false, inputs[i].dtype());
+      auto mem = inputs[i].GetMKLDNNData();
```
Comment (cjolivier01) [inline src/operator/nn/mkldnn/mkldnn_base.cc]: nit: auto should be used for obvious types (ie auto foo = static_cast<Bar *>(bar)) or stuff like iterators which are super-long.  This is just a simnple var.  For readability here, it's be awesome to see what this type actually is.

File: tests/python/gpu/test_gluon_model_zoo_gpu.py (hunk 2 above)
Comment (cjolivier01) [reply tests/python/gpu/test_gluon_model_zoo_gpu.py]: Why would you want to set it the same for each run?

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (marcoabreu) [reply src/ndarray/ndarray.cc]: Yes. Ideally, there should be no TODOs at all. But in cases like this it's acceptable to have them, but they must be tracked as otherwise we lose overview.

**Comment by marcoabreu** [comment]
But would this re-enabled test have caught this error? Please explain how we're preventing this problem from reoccurring.

File: src/operator/nn/mkldnn/mkldnn_base.cc (hunk 4 above)
Comment (zheng-da) [reply src/operator/nn/mkldnn/mkldnn_base.cc]: I saw reserve() being used in many places in MXNet. I'm always wondering how much benefit it brings. Or is this a convention?

**Comment by zheng-da** [comment]
@cjolivier01 The seed is set so we know what is the expected result. It's easier to tell whether CPU or GPU compute wrong results.

**Comment by zheng-da** [comment]
@marcoabreu enabling the tests can catch the error more easily.

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (zheng-da) [reply src/ndarray/ndarray.cc]: The reason I ask is that I saw TODO in many places in the code. Are all TODOs tracked in github issues? Or is this the new practice we want to push? And is github issue the best place for us to keep track of these TODOs? I remember Sheng has a script that closes issues if they aren't inactive for 30 days or so.

**Comment by marcoabreu** [comment]
We don't need fixed seeds since #9791 has been merged. Just make sure that the with_seed decorator is properly applied.

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (marcoabreu) [reply src/ndarray/ndarray.cc]: In think it's best practice and that we should lead by example :) In the last months I haven't seen any TODO (besides the ones in your MKLDNN PR) checked into MXNet. 

We got rid of the script to automatically close issues. For now, GitHub issues are the right place to track these things.

File: tests/python/gpu/test_gluon_model_zoo_gpu.py [hunk 6]
```diff
@@ -37,8 +37,7 @@ def download_data():
     return mx.test_utils.download(
         'http://data.mxnet.io/data/val-5k-256.rec', VAL_DATA)
 
-@unittest.skip("test fails intermittently. temporarily disabled.")
-@with_seed()
+@with_seed(1)
```
Comment (marcoabreu) [inline tests/python/gpu/test_gluon_model_zoo_gpu.py]: Please remove custom seed

File: tests/python/gpu/test_gluon_model_zoo_gpu.py [hunk 7]
```diff
@@ -99,7 +100,7 @@ def get_nn_model(name):
 # Seed 1521019752 produced a failure on the Py2 MKLDNN-GPU CI runner
 # on 2/16/2018 that was not reproducible.  Problem could be timing related or
 # based on non-deterministic algo selection.
-@with_seed()
+@with_seed(1)
```
Comment (marcoabreu) [inline tests/python/gpu/test_gluon_model_zoo_gpu.py]: Please remove custom seed

**Comment by marcoabreu** [comment]
I'm afraid I still don't really understand. I see that you have re-enabled this test, but unfortunately I'm lacking deep technical knowledge of the underlying operations, so would you mind explaining me how this re-enabled test would have caught this failure?

**Comment by zheng-da** [comment]
The reason I disabled the inference tests because I previously thought the failure was related to numeric errors and these are invalid tests. Now it's clear to me that the failures are caused by race condition. It's very hard to reproduce the failure, as you probably have seen. Adding the tests back increases the chance of reproducing the failure if the race condition still exists.

**Comment by marcoabreu** [comment]
Would it be possible to create a test that introduces a delay in order to force this race condition?

File: src/operator/nn/mkldnn/mkldnn_base.cc (hunk 4 above)
Comment (larroy) [reply src/operator/nn/mkldnn/mkldnn_base.cc]: It prevents reallocations, heap fragmentation, and contention if you know the size, it's definitely good to use reserve in advance.

**Comment by zheng-da** [comment]
It's very difficult to reproduce a race condition in a deterministic way if it's possible.

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (larroy) [reply src/ndarray/ndarray.cc]: I'm not familiar with this code. But why are all all these operations mutating the array in a function that is supposed to be const?

File: src/operator/nn/mkldnn/mkldnn_base.cc (hunk 4 above)
Comment (cjolivier01) [reply src/operator/nn/mkldnn/mkldnn_base.cc]: when you add items to a vector, when it runs out of space for the items (which generally happens on the second item sometimes), it allocates memory for 2x the current number of items and then copies all of the items (copy constructor or move, depending on a number of things), into to the new space and then frees the old space. this can happen over and over again, and the whole reallocate and copy can be very expensive. they tend to even show up as hotspots in vtune. so if you know how many items a vector may have, it’s always a good idea to call reserve() because otherwise you are guaranteed a lot of extra reallocing And copying.

**Comment by cjolivier01** [comment]
if you run the test in a loop (from python) 1000 times, does it fail?

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (cjolivier01) [reply src/ndarray/ndarray.cc]: if it doesn’t require const_cast<>, then it’s perfectly legal.

File: src/operator/nn/mkldnn/mkldnn_base.cc (hunk 4 above)
Comment (cjolivier01) [reply src/operator/nn/mkldnn/mkldnn_base.cc]: implementations vary, but this gives a general idea:

http://www.drdobbs.com/c-made-easier-how-vectors-grow/184401375

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (zheng-da) [reply src/ndarray/ndarray.cc]: const can have a lot of definitions. I don't think it's a bad design. here const means the values of the array are always the same, even though its layout may be changed. if we strictly follow the rule that const function doesn't modify data in an array, the code would become unmanagable. In fact, the data in a const array in MXNet is modified all the time.

**Comment by zheng-da** [comment]
it seems the current modification still can't get rid of all race conditions in the code. the reason is that we want to reorder the data of weight arrays in place so that we can avoid reordering them again during inference. Right now, if I ran the tests hundreds of times, I might be able to see inconsistency once. @piiswrong and I have discussed a solution to get rid of remaining race conditions. it should work.

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (larroy) [reply src/ndarray/ndarray.cc]: Why don't we use dynamic dispatch in NDArray for different implementations and storage options? it would make the code more mainable. I predict Chris is going to mention the overhead of a virtual function but I don't think it would be significant also considering the conditional logic in some of the NDArray methods and Python overheads which are magnitudes higher.

**Comment by cjolivier01** [comment]
one thing i’ve done in the past to try and make race conditions happen more often is start changing the process affinity mask, like for example, force the process to use one or two cpus only. this is separate from OMP.

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (larroy) [reply src/ndarray/ndarray.cc]: @cjolivier01  sure it's legal to change data through a pointer from a const method, but having unexpected side effects from a class contract which indicates that a method is const leads to problems, is not particularly considered a good practice when programming: https://stackoverflow.com/questions/763835/are-side-effects-a-good-thing  having experience with functional programming languages helps appreciate controlled side effects.

Comment (cjolivier01) [reply src/ndarray/ndarray.cc]: There are many decaffeinated brands out there that are just as tasty as the real thing.

**Comment by larroy** [comment]
Is this need to reorder data documented somewhere?

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (marcoabreu) [reply src/ndarray/ndarray.cc]: Whatever you guys are talking here... I think it's a pretty basic thing that immutable means immutable and that a getter should never modify the underlying data in any way. I have no clue about C++, but I'm quite certain that this constraint is quite clearly defined across the industry. Could we please get back to a productive discussion?

Comment (zheng-da) [reply src/ndarray/ndarray.cc]: @larroy I agree with you that the current implementation of NDArray to handle storages is messy. What you said requires a rewrite of NDArray. @piiswrong and I talked about this, but I don't think this should be in this PR. We need to be very careful when rewriting NDArray, considering NDArray is one of the core components in MXNet.

**Comment by zheng-da** [comment]
@cjolivier01 why race condition happens more frequently when threads run in a smaller number of CPU cores? It seems to me that it happens more often in more CPU cores.

**Comment by zheng-da** [comment]
@larroy not yet. this is the design doc of mkldnn: https://cwiki.apache.org/confluence/display/MXNET/The+design+of+MKLDNN+integration
I'll add it to the doc.

**Comment by cjolivier01** [comment]
@zheng-da Well, you can vary the cores up or down just to change the timing and try to make the race condition happen more often by changing the timing to a point where it's especially aggravated.

I have found, however, that reducing the number of cores that actually run the program (not the number of threads) can cause more overlap between threads.  By this, I mean, each thread gets a smaller timeslice of actual execution, and so things that might "overlap" tend to overlap with more frequency.  Think of two trains passing each other in opposite directions (threads).  If the trains are going fast, they are only next to each other for a short period of time.  But if the two trains are going slowly, then they are next to each other for a much longer period of time.

**Comment by cjolivier01** [comment]
That's not to say the same trick works for any race condition (fewer cores). But varying the processor affinity in either direction can help to flush out race conditions in general.

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (larroy) [reply src/ndarray/ndarray.cc]: @zheng-da I didn't mean it to be included in this PR. I was just asking. Anyway thanks for your responses. I didn't get the quote about the brands.

**Comment by TaoLv** [comment]
@zheng-da, if needed, please also update the design document for mkldnn integration. Thanks.

**Comment by zheng-da** [comment]
@TaoLv I have updated the design doc to explain why we need data layout conversion.

File: src/ndarray/ndarray.cc [hunk 8]
```diff
@@ -375,7 +375,45 @@ void NDArray::Chunk::Reorder2Default() {
...
+  CHECK(old_mem->get_primitive_desc().desc().data.ndims == _desc.data.ndims);
+
+  // This may be called in MKLDNN operators. We can't use MKLDNNStream here.
+  std::vector<mkldnn::primitive> net;
+  net.push_back(mkldnn::reorder(*old_mem, *new_mem));
+  mkldnn::stream(mkldnn::stream::kind::eager).submit(net).wait();
+
+  CHECK(shandle.size >= pd.get_size());
+  CheckAndAlloc(pd.get_size());
+  // TODO(zhengda) We need to avoid memory copy here.
```
Comment (marcoabreu) [inline src/ndarray/ndarray.cc]: Open TODO

File: src/ndarray/ndarray.cc (hunk 1 above)
Comment (marcoabreu) [reply src/ndarray/ndarray.cc]: Open TODO

File: src/ndarray/ndarray.cc (hunk 8 above)
Comment (zheng-da) [reply src/ndarray/ndarray.cc]: This is from the previous PR. I just moved code here.

**Comment by zheng-da** [comment]
@marcoabreu Reorder2Default and MKLDNNDataReorder shouldn't be called frequently. They are not in the critical path. The whole point of this PR is to further remove the invocation of these two methods.

**Comment by pengzhao-intel** [comment]
@zheng-da @marcoabreu @cjolivier01 please since there're lots of changes in MKL-DNN, please kindly wait a moment to merge.

@juliusshufan will test the performance and coverage of these changes.

**Comment by zheng-da** [comment]
@cjolivier01 do you have more comments?
@piiswrong do you want to review the code?

The PR should have fixed the bug in https://github.com/apache/incubator-mxnet/issues/9820.
I ran tests/python/gpu/test_gluon_model_zoo_gpu.py again 1000 times and didn't see any failure. I also ran tests/python/gpu/test_operator_gpu.py many times (~40 times so far) and saw two failures, but none of them are caused by this race condition. The failures in test_operator_gpu.py are caused by floating-point precision. I'll report them in another issue.

**Comment by piiswrong** [comment]
LGTM.
@pengzhao-intel have you finished testing?

**Comment by cjolivier01** [comment]
LGTM

**Comment by juliusshufan** [comment]
@zheng-da @cjolivier01 @piiswrong @zheng-da sorry for late response.
The test result is positive, with 300 epoc, resnet50+cifar10
top1 validation accuracy is 0.923478
top1 training accuracy is 0.987500

Thanks.

**Comment by pengzhao-intel** [comment]
The testing is done. There is **NO** coverage and performance regression issue from our tests.
I think the code is **qualifie**d to merge @piiswrong  :) 

@juliusshufan will update the performance data in here soon.

Thank you for @zheng-da's great work.

**Comment by marcoabreu** [comment]
Thanks a lot everybody!

**Comment by juliusshufan** [comment]
Update the inference perf comparison, a slight drop on ResNet network, no outstanding changes on other networks. 

  | Batch Size | with Da's fix | w.o Da's fix
-- | -- | -- | --

AlexNet | 1 | 347.56 | 346.82
128 | 1910.29 | 1911.13
256 | 1958.21 | 1928.43
VGG-16 | 1 | 99.1 | 101.07
128 | 133.12 | 134.95
256 | 132.77 | 132.55
ResNet50 | 1 | 79.11 | 82.27
128 | 191.71 | 191.48
256 | 189.55 | 186.17
Inception-v3 | 1 | 53.19 | 54.86
128 | 218.46 | 218.96
256 | 217.85 | 213.03
ResNet-152 | 1 | 31.21 | 33.78
128 | 95.02 | 81.59
256 | 90.87 | 77.69
Inception-bn | 1 | 102.04 | 102.22
128 | 564.2 | 564.32
256 | 533.87 | 534.40
